pkg/virt-controller/watch/export/export.go

```diff
@@ -384,19 +393,143 @@ func (ctrl *VMExportController) handleIsSourcePvc(vmExport *exportv1.VirtualMach
+			}
+		}
+	} else {
+		if err := ctrl.handlePodSucceededOrFailed(vmExport, pod); err != nil {
+			return retry, err
 		}
 	}
 
-	return ctrl.updateVMExportPvcStatus(vmExport, pvcs, pod, service, inUse)
+	return ctrl.updateVMExportPvcStatus(vmExport, pvcs, pod, service, false)
+}
+
+func (ctrl *VMExportController) handlePodSucceededOrFailed(vmExport *exportv1.VirtualMachineExport, pod *corev1.Pod) error {
+	if pod.Status.Phase == corev1.PodSucceeded || pod.Status.Phase == corev1.PodFailed {
+		// The server died or completed, delete the pod.
+		ctrl.Recorder.Eventf(vmExport, corev1.EventTypeWarning, exporterPodFailedOrCompletedEvent, "Exporter pod %s/%s is in phase %s", pod.Namespace, pod.Name, &pod.Status.Phase)
+		return ctrl.Client.CoreV1().Pods(vmExport.Namespace).Delete(context.Background(), pod.Name, metav1.DeleteOptions{})
+	}
+	return nil
+}
+
+func (ctrl *VMExportController) handlePVCsForVirtualMachineSnapshot(vmExport *exportv1.VirtualMachineExport, vmSnapshot *snapshotv1.VirtualMachineSnapshot) ([]*corev1.PersistentVolumeClaim, int, error) {
+	var content *snapshotv1.VirtualMachineSnapshotContent
+	var err error
+	pvcs := make([]*corev1.PersistentVolumeClaim, 0)
+	exists := false
+	totalVolumes := 0
+
+	if vmSnapshot.Status.VirtualMachineSnapshotContentName != nil && *vmSnapshot.Status.VirtualMachineSnapshotContentName != "" {
+		content, exists, err = ctrl.getVmSnapshotContent(vmSnapshot.Namespace, *vmSnapshot.Status.VirtualMachineSnapshotContentName)
+		if err != nil {
+			return pvcs, totalVolumes, err
+		} else if exists {
+			sourceVm := content.Spec.Source.VirtualMachine
+			totalVolumes = len(content.Status.VolumeSnapshotStatus)
+			for _, volumeBackup := range content.Spec.VolumeBackups {
+				if pvc, err := ctrl.getOrCreatePVCFromSnapshot(vmExport, &volumeBackup, sourceVm); err != nil {
+					return pvcs, totalVolumes, err
+				} else if pvc != nil {
+					pvcs = append(pvcs, pvc)
+				}
+			}
+		}
+	}
+	return pvcs, totalVolumes, err
+}
+
+func (ctrl *VMExportController) getOrCreatePVCFromSnapshot(vmExport *exportv1.VirtualMachineExport, volumeBackup *snapshotv1.VolumeBackup, sourceVm *snapshotv1.VirtualMachine) (*corev1.PersistentVolumeClaim, error) {
+	if volumeBackup.VolumeSnapshotName == nil {
+		log.Log.Errorf("VolumeSnapshot name missing %+v", volumeBackup)
+		return nil, fmt.Errorf("missing VolumeSnapshot name")
+	}
+	restorePVCName := fmt.Sprintf("%s-%s", vmExport.Name, volumeBackup.PersistentVolumeClaim.Name)
+
+	if pvc, exists, err := ctrl.getPvc(vmExport.Namespace, restorePVCName); err != nil {
+		return nil, err
+	} else if exists {
+		return pvc, nil
+	}
+
+	volumeSnapshot, err := ctrl.VolumeSnapshotProvider.GetVolumeSnapshot(vmExport.Namespace, *volumeBackup.VolumeSnapshotName)
+	if err != nil {
+		return nil, err
+	}
+
+	// leaving source name and namespace blank because we don't care in this context
+	pvc := snapshot.CreateRestorePVCDef("", restorePVCName, volumeSnapshot, volumeBackup, "", "")
```

Comment:
Can this be refactored to not take the empty string args?  Instead of having one function that is shared with restore, have two (or more)?

Reply:
Sure